### PR TITLE
Upgrade pre-commit configuration and move to ruff

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: doc/conf.py [hunk 1]
```diff
@@ -62,7 +62,7 @@
 # built documents.
 #
 # The short X.Y version.
-from astroid.__pkginfo__ import __version__
+from astroid.__pkginfo__ import __version__  # noqa
```
Comment (DanielNoord) [inline doc/conf.py:65]: Why is this necessary?

File: pyproject.toml [hunk 2]
```diff
@@ -65,14 +65,6 @@ python_files = ["*test_*.py"]
 testpaths = ["tests"]
 filterwarnings = "error"
```
Comment (DanielNoord) [inline pyproject.toml:67]: Is the `isort` config no longer neede?

File: requirements_test_pre_commit.txt [hunk 3]
```diff
@@ -1,9 +1,5 @@
 -r requirements_test_min.txt
 black==23.1.0
```
Comment (DanielNoord) [inline requirements_test_pre_commit.txt:2]: I'd be okay with removing this as well

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 4]
```diff
@@ -2927,7 +2926,7 @@ def _slots(self):
             if exc.args and exc.args[0] not in ("", None):
                 return exc.args[0]
             return None
-        return [first] + list(slots)
+        return [first, *list(slots)]
```
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:2929]:
```suggestion
        return [first, *slots]
```

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 5]
```diff
@@ -3033,7 +3032,7 @@ def _compute_mro(self, context: InferenceContext | None = None):
                 ancestors = list(base.ancestors(context=context))
                 bases_mro.append(ancestors)
 
-        unmerged_mro = [[self]] + bases_mro + [inferred_bases]
+        unmerged_mro = [[self], *bases_mro] + [inferred_bases]
```
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes/scoped_nodes.py:3035]:
```suggestion
        unmerged_mro = [[self], *bases_mro, inferred_bases]
```

File: pyproject.toml (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pyproject.toml:67]: I don't think ruff is using it, it's compatible with black by default.

File: requirements_test_pre_commit.txt (hunk 3 above)
Comment (Pierre-Sassoulas) [reply requirements_test_pre_commit.txt:2]: Removing black or removing the version ?

File: tests/test_regrtest.py [hunk 6]
```diff
@@ -264,7 +264,7 @@ def test_unicode_in_docstring(self) -> None:
 
         class MyClass(object):
             def method(self):
-                "With unicode : {'’'} "
+                "With unicode : {'`'} "
```
Comment (Pierre-Sassoulas) [inline tests/test_regrtest.py:267]: This looks suspicious, having strange unicode characters is probably what we want to check.

File: pyproject.toml (hunk 2 above)
Comment (DanielNoord) [reply pyproject.toml:67]: But do we get import sorting still then?

File: requirements_test_pre_commit.txt (hunk 3 above)
Comment (DanielNoord) [reply requirements_test_pre_commit.txt:2]: versioN!

File: pyproject.toml (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pyproject.toml:67]: We might not have intra group sorting with first / third parties, which is why isort is kept in pylint. I think in astroid we'll have "good enough sorting" because we don't have a lot of third parties -- and faster commits.

Comment (Pierre-Sassoulas) [reply pyproject.toml:67]: Nevermind, I just tested, ruff sort correctly first party and third party without any configuration.

Comment (Pierre-Sassoulas) [reply pyproject.toml:67]: We might want to use ruff instead of isort for wrong-import-order in pylint.

File: requirements_test_pre_commit.txt [hunk 7]
```diff
@@ -1,9 +1,5 @@
...
-isort==5.12.0;python_version>='3.8'
-flake8==6.0.0;python_version>='3.8'
-flake8-typing-imports==1.14.0;python_version>='3.8'
-flake8-bugbear==23.2.13;python_version>='3.8'
-mypy==1.1.1
-pre-commit~=2.21
+black
+pylint
+mypy
+pre-commit
```
Comment (Pierre-Sassoulas) [inline requirements_test_pre_commit.txt:5]:
```suggestion
ruff
pre-commit
```

**Review by DanielNoord — APPROVED** [review]
